### DISCUSSION TITLE
Shortcut for addition of DYLIB_API before functions and variables

### DISCUSSION THREAD

**Discussion by sethupavan12** [martin-olivier/dylib#11] [Q&A] [answered]
Hi there,

I came across dylib a couple of days ago and I thought it could be great tool.  

The way I am currently creating a DLL/.so files is through compiling .m (matlab) files through MATLAB Coder which converts Matlab to MEX(Matlab Executable) or C++ files. There can potentially be lots of varaibles/ functions genereated by Coder during compilation and it would be really hard to put a DYLIB_API infront of them manually since they are intially in MATLAB. 

Is there any alterantive to this? Perhaps something of inheritence nature?
PS:

**Comment by martin-olivier** [comment, marked as answer]
Hello,
The goal of `DYLIB_API` is to mark a function or a variable as `extern "C"` and put `dllexport` on windows.

https://www.mathworks.com/help/compiler_sdk/gs/create-a-c-application-with-matlab-code.html

In Matlab, if you build your code as a **C** dynamic library, the usage of `extern "C"` is now useless
On the link above, I saw that you have the option "exported function", this option will add `dllexport` before your exported function

With that, you should no longer need to add manually `DYLIB_API`.
After building into a C dynamic library, you can check if your symbols are here and correctly named using `nm`:
https://en.wikipedia.org/wiki/Nm_(Unix)

**Reply by sethupavan12** [reply]
Thank you so much for the clear explanation. 
There's just one thing I'd like to know.
So, does this mean I could just make a C dynamic library using matlab (with no  `DYLIB_API` ) and use `dylib lib("./cDynamicLib", dylib::extension);` to load the dynamic lib and use ` .getfunction() ` to get a function directly?

**Reply by martin-olivier** [reply]
What you said will work for sure on unix os (linux, macOS).
But on windows, the functions you want to load dynamically **MUST** be marked as `__declspec(dllexport)`.
For that, you will need to use the option `Exported Functions` in mathlab when you are generating your dynamic library like shown here:
https://www.mathworks.com/help/compiler_sdk/gs/create-a-c-application-with-matlab-code.html

**Reply by sethupavan12** [reply]
Great. 
Thank you, Martin!

**Reply by sethupavan12** [reply]
Just to try out the suggested way ....
I tried generating a C dynamic lib using the libray compiler app in MATLAB and tried to extract a function from the generated DLL.
.m -> DLL -> extract usign DYLIB

Here is the matlab file that has been compiled into a DLL
```
function output = compile(a)
    % find nth prime number using the sieve of Eratosthenes and store in output
    sieve = ones(1,a);
    sieve(1) = 0;
    sieve(2) = 0;
    % store the primes in output
    output = [];
    for i = 1:a
        if sieve(i) == 1
            output = [output i];
        end
    end

end
```
 
and when I load the DLL using a C++ file (main.cpp) I got the following error
```
dylib: error while loading symbol "compile"
The specified procedure could not be found. 
```

main.cpp which is modified version of your example in readme.md
```
#include <iostream>
#include "dylib.hpp"

int main()
{
    try {
        dylib lib("./compileLibCompiler", dylib::extension);

        auto nth_prime = lib.get_function<int(int)>("compile");
        //nth_prime(5);
        int a = 5;
        int ans = nth_prime(a);
        printf("%d\n", ans);

        return 0;

     }
    catch (const dylib::exception &e) {
        std::cerr << e.what() << std::endl;
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}
```

What do you think went wrong here?

**Reply by martin-olivier** [reply]
Hello, do you have used the option `Exported Functions` while you were generating your dynamic library ?
If it's the case can you check if the symbols are here and correctly named on your dynamic library by using `dumpbin`

**Reply by martin-olivier** [reply]
There is also this tool if you want to see witch symbols are exported and see their name :
https://www.nirsoft.net/utils/dll_export_viewer.html

**Reply by sethupavan12** [reply]
Hi @martin-olivier , I did select the option `Export Functions` so I can not worry about __dllspec stuff as u said. The problem I see here with this is that when I see the dll using the above software, I can no longer see a function named compile. it is changed to the following. 

initially in compile.m function was named "compile" 

after library compiler dll generation, 
```

compileInitialize	0x0000000180001030	0x00001030	1 (0x1)	compile.dll	Exported Function	
compileInitializeWithHandlers  0x00000001800010f0	0x000010f0	2 (0x2)  compile.dll	Exported Function	
compilePrintStackTrace	0x00000001800011a0	0x000011a0	3 (0x3)	compile.dl	Exported Function	
compileTerminate	0x0000000180001230	0x00001230	4 (0x4)	compile.dll	Exported Function	
mlfCompile	0x00000001800012f0	0x000012f0	5 (0x5)	compile.dll      Exported Function	
mlxCompile	0x0000000180001370	0x00001370	7 (0x7)	compile.dll	        Exported Function	

	
```
If I just take some random function here, I wouldn't know what they return or what they actually do.

**Reply by martin-olivier** [reply]
Hey, sorry I don't have much knowledge about Matlab, I found this guide to load dynamic libs generated from Matlab, and as you can see the process seems quite different compared to native C/C++ shared libs: (He calls the symbol `libnameInitialise`, `libnameTerminate`, ...).
I think you will have to do a little wrapper around `dylib` in order to achieve your project 👍 
https://www.mathworks.com/help/ecoder/ug/creating-and-using-host-based-shared-libraries.html

**Reply by martin-olivier** [reply]
Another option will be to generate a C code from your .m
then build your C code into a dynamic library with `CMake` using this flag:
```cmake
if(MSVC)
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()
```
This flag will export all symbols on windows without you needing to mark each function as `__declspec(dllexport)`

Full example:

First generate your matlab code (.m) into a C code (code.c)
Then create a file named `CMakeLists.txt` and fill this file with those instructions:
```cmake
cmake_minimum_required(VERSION 3.14)
project(matlab_loader CXX)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR})
if(MSVC)
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

add_library(matlab_lib SHARED code.cpp)
set_target_properties(matlab_lib PROPERTIES PREFIX "")
```

then open a shell and type the following commands:
```sh
cmake . -B build
cmake --build build
```

After that you will find `matlab_lib.dll` on the same folder as your `CMakeLists.txt` or in `Debug` folder.
After that you will be able to load your symbols with `dylib` without issues

**Reply by sethupavan12** [reply]
Yes! I was thinking about Cmake in a diff way but this is much simpler. Especially with that flag, u mentioned. You are a legend!
I will try it out now.

**Reply by martin-olivier** [reply]
There is an error on the `CMakeLists` i just send, the `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` must be set before creating the dynamic lib 😅 
```cmake
cmake_minimum_required(VERSION 3.14)
project(matlab_loader CXX)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR})
if(MSVC)
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

add_library(matlab_lib SHARED code.cpp)
set_target_properties(matlab_lib PROPERTIES PREFIX "")
```

**Reply by sethupavan12** [reply]
haha no problem! Thanks

Do you think I can do the same thing for .so generation on linux?

**Reply by martin-olivier** [reply]
Yes it will work for `Windows`, `MacOS` and `Linux`.
You will just need to use `dylib::extension` while loading your lib 👍

**Reply by sethupavan12** [reply]
Oh, I forgot to mention. The C code is generated using Matlab coder which generates completely random internal functions but I was able to generate and build a DLL. but it is just full of random functions it generates internally.
I guess I would have to play around to make sure it doesn't destroy the original function name in the settings or something.

Another option I have is to manually convert .m to C to build a DLL
I will try a few things as well with coder settings
Thank u so much for your help man. It helped a lot! 

PS: I will keep you updated.

**Reply by martin-olivier** [reply]
Thank you 👍 
On those examples, the function name seems to be properly keeped during the conversion from .m to c code:
https://www.youtube.com/watch?v=-rYfvAEkY5g
https://fr.mathworks.com/help/coder/gs/generating-c-code-from-matlab-code-using-the-matlab-coder-project-interface.html
Keep me updated :)

**Reply by sethupavan12** [reply]
So, I've basically just converted the .m to .cpp using an option called "Source Code Generation". Perhaps I'll try the library mode they have. Thanks for the example 😎

**Reply by martin-olivier** [reply]
I think i know why you have random internal functions names.
C++ uses name mangling while building libraries or source code to be able to differentiate function `overloads` or functions with the same name on different namespaces:
https://en.wikipedia.org/wiki/Name_mangling
If you build your code into **C** source code it should fix the problem

**Reply by sethupavan12** [reply]
Yes. I think I did about name mangling somewhere but didn't know coder used it.

Will CMAKE be happy if I build C source code into a DLL?

I accidentally compiled into C but then I discarded it thinking Cmake wouldn't be happy and I just wanted to get something working first.

**Reply by martin-olivier** [reply]
yes, my bad, replace your `CMakeLists.txt` with this content, it will work for **C** code, assuming that your c source code generated from matlab is named `code.c`:
```cmake
cmake_minimum_required(VERSION 3.14)
project(matlab_loader C)

set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR})
if(MSVC)
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

add_library(matlab_lib SHARED code.c)
set_target_properties(matlab_lib PROPERTIES PREFIX "")
```

```c
// code.c

int adder(int a, int b)
{
    return a + b;
}
```
Remove the `build` folder if you already have one and then enter in a shell:
```
cmake . -B build
cmake --build build
```

After that, the symbol is correctly named:
```
➜ nm ./matlab_lib.so
0000000000003fa0 T adder
```

**Reply by sethupavan12** [reply]
Hi Martin, I've tried doing this with C but I think Coder has an internal way of "generating predefined types and linking it to the Matlab-C APIs"
So, after generation, I built it and when `cmake --build build` it I get the following unresolved issues
```
compile.obj : error LNK2019: unresolved external symbol emxEnsureCapacity_int32_T referenced in function compile [C:\Users\\Documents\toy\matlab_DLL_extraction\compile_pkg3\build\Cdll.vcxproj]
compile.obj : error LNK2019: unresolved external symbol emxEnsureCapacity_int8_T referenced in function compile [C:\Users\\Documents\toy\matlab_DLL_extraction\compile_pkg3\build\Cdll.vcxproj]
compile.obj : error LNK2019: unresolved external symbol emxFree_int8_T referenced in function compile [C:\Users\\Documents\toy\matlab_DLL_extraction\compile_pkg3\build\Cdll.vcxproj]
compile.obj : error LNK2019: unresolved external symbol emxInit_int8_T referenced in function compile [C:\Users\\Documents\toy\matlab_DLL_extraction\compile_pkg3\build\Cdll.vcxproj]

```
When I convert the matlab to C the following files are generated for compile.m
```
03/03/2022  10:38           217,912 buildInfo.mat
03/03/2022  10:41               276 CMakeLists.txt
03/03/2022  10:38             1,596 compile.c
03/03/2022  10:38               706 compile.h
03/03/2022  10:38               439 compile_data.c
03/03/2022  10:38               535 compile_data.h
03/03/2022  10:38             3,056 compile_emxAPI.c
03/03/2022  10:38             1,304 compile_emxAPI.h
03/03/2022  10:38             4,615 compile_emxutil.c
03/03/2022  10:38             1,129 compile_emxutil.h
03/03/2022  10:38               581 compile_initialize.c
03/03/2022  10:38               706 compile_initialize.h
03/03/2022  10:38               615 compile_terminate.c
03/03/2022  10:38               701 compile_terminate.h
03/03/2022  10:38             1,306 compile_types.h
03/03/2022  10:38                41 defines.txt
03/03/2022  10:38             1,257 rtwtypes.h
03/03/2022  10:38               135 rtw_proj.tmw
15/11/2020  00:18            22,875 tmwtypes.h
03/03/2022  10:38               696 _clang-format
```


Surprisingly, CMake build doesn't seem to fail with cpp 😢

**Reply by sethupavan12** [reply]
Also a small question,

```
if(MSVC)
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
endif()

```

this is exclusive to windows right..
what if I am on linux, do I need to export something?

**Reply by martin-olivier** [reply]
Hey,

On linux and macOS, all symbols are exported by default, the cmake flag `EXPORT_ALL_SYMBOLS` is only necessary on Windows.
Can you send me the files that matlab generates when you convert `compile.m` to C code ? (wetransfer)

**Reply by sethupavan12** [reply]
Great. Cheers

Here is the link https://we.tl/t-2GeAVc2IVP
I've never used this before seems like good service. Ignore the LinkedIn message

**Reply by martin-olivier** [reply]
Hello Seth,
I sent you a repository on github that contains the two parts of the process :
- The lib factory
- The lib loader

You will need to put the following commands :
```c++
cmake . -B build
cmake --build build
cd Debug
./MatlabLoader
```
It worked on linux too

Actually the code is crashing inside the `compile` function, if you put some `printf` inside the function, it loops 3 times and crashes on linux (perhaps because missing allocations inside the struct `emxArray_int32_T` passed as parameter)

Hope it helped :)

**Reply by sethupavan12** [reply]
Amazing! 
Thanks again, Martin. I will try it out and let you know

**Reply by sethupavan12** [reply]
Hey, thanks for the help man.

So the workaround is basically to make sure we pass all the parameters that are in generated C function while using it right?

I think I got enough ground to integrate this. Couldn't have without you thanks again mate ❤️
